packages/office-ui-fabric-react/src/components/ComboBox/ComboBox.tsx

```diff
@@ -1241,6 +1251,56 @@ export class ComboBox extends BaseComponent<IComboBoxProps, IComboBoxState> {
     }
   }
 
+  private _notifyPreviewChanged(prevState: IComboBoxState) {
+    const { onPreviewExecute, onRevertPreviewExecute } = this.props;
+    const {
+      currentPendingValue,
+      currentOptions,
+      currentPendingValueValidIndex,
+      currentPendingValueValidIndexOnHover
+    } = this.state;
+
+    if (onPreviewExecute && onRevertPreviewExecute) {
+      let sendRevert = false;
+      let currentPendingIndex: number | undefined = undefined;
+      let pendingValue: string | undefined = undefined;
+
+      if (currentPendingValueValidIndexOnHover !== prevState.currentPendingValueValidIndexOnHover) {
+        if (this._indexWithinBounds(currentOptions, currentPendingValueValidIndexOnHover)) {
+          currentPendingIndex = currentPendingValueValidIndexOnHover;
+        } else {
+          sendRevert = true;
+        }
+      }
+
+      if (!currentPendingIndex && currentPendingValueValidIndex !== prevState.currentPendingValueValidIndex) {
+        if (this._indexWithinBounds(currentOptions, currentPendingValueValidIndex)) {
+          currentPendingIndex = currentPendingValueValidIndex;
+        } else {
+          sendRevert = true;
+        }
+      }
+
+      if (!currentPendingIndex && currentPendingValue !== prevState.currentPendingValue) {
+        if (currentPendingValue !== '') {
+          pendingValue = currentPendingValue;
+        } else {
+          sendRevert = true;
+        }
+      }
+
+      if (this._isInPreview && (sendRevert || currentPendingIndex || pendingValue)) {
+        onRevertPreviewExecute();
+        this._isInPreview = false;
+      }
+
+      if (!this._isInPreview && (currentPendingIndex || pendingValue)) {
+        onPreviewExecute(currentPendingIndex ? currentOptions[currentPendingIndex] : undefined, currentPendingIndex, pendingValue);
+        this._isInPreview = true;
+      }
+    }
```

Comment:
Can you add comments for all of these if checks? This seems overly complex and it could potentially be restructured to be more simple and straight forward.